Fix 24073 - validate date and handle time offset

ArwenQin opened the pull request:
*Issue #:* /bcgov/entity#24073

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

severinbeauvais commented on legal-api/src/legal_api/services/filings/validations/continuation_in.py:
```diff
@@ -200,13 +201,10 @@ def validate_business_in_colin(filing_json: dict, filing_type: str) -> list:
             else:
-                if dt.fromisoformat(founding_date) != dt.fromisoformat(response_json['business']['foundingDate']):
+                utc = pytz.timezone('UTC')
+                if dt.fromisoformat(founding_date).astimezone(utc).date() != dt.fromisoformat(response_json['business']['foundingDate']).astimezone(utc).date():
```
Why do you need `astimezone()`?

ArwenQin replied:
I'm thinking about handling the timezone offset like '2007-08-19T01:03:18+03:00' is actually Aug 18 in UTC

ArwenQin replied:
If date and time from COLIN is always in UTC, I think we can just compare 
if dt.fromisoformat(founding_date).date() != dt.fromisoformat(response_json['business']['foundingDate']).date()

vysakh-menon-aot replied:
Does UI change the value received from colin (I was expecting UI is passing what colin-api provides)?

vysakh-menon-aot replied:
colin-api provides utc ref: https://github.com/bcgov/lear/blob/main/colin-api/src/colin_api/utils/__init__.py#L49

ArwenQin replied:
Yes, this is in create UI;
![image](https://github.com/user-attachments/assets/546c25a1-dbd3-4565-a279-40a800e86052)

![image](https://github.com/user-attachments/assets/3248e546-1923-4ddc-9340-7a92c615f913)


![image](https://github.com/user-attachments/assets/288c0ecb-2e97-4685-afa6-d51684412151)

ArwenQin replied:
Thus, I think we can just compare: 
dt.fromisoformat(founding_date).date() != dt.fromisoformat(response_json['business']['foundingDate']).date()

(just compare the date, ignore the time)

severinbeauvais replied:
> I'm thinking about handling the timezone offset like '2007-08-19T01:03:18+03:00' is actually Aug 18 in UTC

No. UTC date is Aug 19. But Pacific date is Aug 18.

severinbeauvais replied:
Vysakh, yes, the UI converts the date-time from COLIN into a date-only. This is so that the user can override it if needed:

![image](https://github.com/user-attachments/assets/bebf53c9-6705-41a0-b5c5-9991fcfc0c74)

Upon filing, the UI converts the date into ISO format (ie, time offset is 0700 or 0800).

severinbeauvais replied:
> If date and time from COLIN is always in UTC, I think we can just compare if dt.fromisoformat(founding_date).date() != dt.fromisoformat(response_json['business']['foundingDate']).date()

This looks good to me.

vysakh-menon-aot replied:
> > Does UI change the value received from colin (UI should be passing what colin-api provides)?
> 
> Vysakh, yes, the UI converts the date-time from COLIN into a date-only. This is so that the user can override it if needed:
> 
> ![image](https://private-user-images.githubusercontent.com/10002889/381706373-bebf53c9-6705-41a0-b5c5-9991fcfc0c74.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAzMTc3NDYsIm5iZiI6MTczMDMxNzQ0NiwicGF0aCI6Ii8xMDAwMjg4OS8zODE3MDYzNzMtYmViZjUzYzktNjcwNS00MWEwLWI1YzUtOTk5MWZjZmMwYzc0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDMwVDE5NDQwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE4MTdmYmQyYzFiNjYwYjI5MTBhN2RjMGQzMWRmZjAxNWU3ZTQ5NmMzODUxNzZhNzUwMTg0MzA5N2EyODg5MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.MZHOjjkT9dNLNQ6jFlfovs1hQ8JI9f0TnF6OfYxt4ZU)
> 
> Upon filing, the UI converts the date into ISO format (ie, time offset is 0700 or 0800).
> 
> So the original time is not retained.

User can override `homeRecognitionDate` not `foundingDate`

severinbeauvais replied:
Oh, dang it. I keep getting confused.

So maybe there's a Create UI bug ticket in there somewhere. Arwen, let me know.

ArwenQin replied:
> Oh, dang it. I keep getting confused.
> 
> So maybe there's a Create UI bug ticket in there somewhere. Arwen, let me know.

This is in create UI, that's why the format is changed
![image](https://github.com/user-attachments/assets/6a509d22-f52f-4e9b-8119-2007b02b1aee)


![image](https://github.com/user-attachments/assets/1f6d6c8e-baea-4590-a5ac-614beae0c282)

![image](https://github.com/user-attachments/assets/280c1702-d8fb-434f-a401-ca3eba6685d4)

ArwenQin replied:
Is it this ticket:
https://app.zenhub.com/workspaces/entities---olga-65af15f59e89f5043c2911f7/issues/gh/bcgov/entity/21135

severinbeauvais replied:
No, I meant that we might need a new ticket.

OK, so Create UI does indeed convert the Founding Date from COLIN into a YYYY-MM-DD. Since it looks like the user cannot update this date (ie, it's read only) then we should change the code to retain the time (and convert to date-only where needed).

Arwen, do you agree? If so, please create a new ticket and can you work on it today?

As for the Legal API code, does it need to be changed after all?

ArwenQin replied:
Yes I agree. I just created #24141, I will work on it today.
After fixing it in the create UI, I will update Legal API codes to just compare the string

severinbeauvais replied:
How about we don't merge this PR and just merge your changes later?

ArwenQin replied:
Sure, I agree, I will push changes to the current PR later. We can review and merge it then.

severinbeauvais reviewed (commented):
~Can you update your unit tests to have a COLIN date with time?~

_I see you've now done this. Thanks!_

severinbeauvais commented on legal-api/src/legal_api/services/filings/validations/continuation_in.py:
```diff
@@ -200,13 +201,10 @@ def validate_business_in_colin(filing_json: dict, filing_type: str) -> list:
-                        'error': (
-                            f"Founding date does not match with founding date from Colin. "
-                            f"Filing founding_date: {founding_date}, "
-                            f"Colin API founding_date: {response_json['business']['foundingDate']}"
```
Yes, thanks for removing the debugging code.

severinbeauvais commented on legal-api/src/legal_api/services/filings/validations/continuation_in.py:
```diff
@@ -192,7 +192,7 @@ def validate_business_in_colin(filing_json: dict, filing_type: str) -> list:
             msg.append({'error': 'Legal name does not match with company legal name from Colin.',
                         'path': business_legal_name_path})
-        elif founding_date != response_json['business']['founding_date']:
+        elif founding_date != response_json['business']['foundingDate']:
```
Did the unit test catch this typo?

ArwenQin replied:
Yes. I noticed this typo when I saw the failed unit test.

ArwenQin commented:
@vysakh-menon-aot @severinbeauvais 
Since we updated the create-ui to retain the founding date fetched from Colin, I changed the legal-api codes back to just compare the string.
Could you please merge it, thank you!
